00X-ABI-Stable-Module-API.md

```diff
@@ -0,0 +1,377 @@
+// Methods to support ObjectWrap
+NODE_EXTERN napi_value napi_create_constructor_for_wrap(napi_env e,
+                                                        napi_callback cb);
+NODE_EXTERN void napi_wrap(napi_env e, napi_value jsObject, void* nativeObj,
+                           napi_destruct* napi_destructor,
+                           napi_persistent* handle);
+NODE_EXTERN void* napi_unwrap(napi_env e, napi_value jsObject);
+
+NODE_EXTERN napi_value napi_create_constructor_for_wrap_with_methods(
+  napi_env e,
+  napi_callback cb,
+  char* utf8name,
+  int methodcount,
+  napi_method_descriptor* methods);
+
+// Methods to control object lifespan
+NODE_EXTERN napi_persistent napi_create_persistent(napi_env e, napi_value v);
+NODE_EXTERN void napi_release_persistent(napi_env e, napi_persistent p);
+NODE_EXTERN napi_value napi_get_persistent_value(napi_env e, napi_persistent p);
+NODE_EXTERN napi_handle_scope napi_open_handle_scope(napi_env e);
+NODE_EXTERN void napi_close_handle_scope(napi_env e, napi_handle_scope s);
+NODE_EXTERN napi_escapable_handle_scope
+                napi_open_escapable_handle_scope(napi_env e);
+NODE_EXTERN void napi_close_escapable_handle_scope(
+                                             napi_env e,
+                                             napi_escapable_handle_scope s);
+NODE_EXTERN napi_value napi_escape_handle(napi_env e,
+                                          napi_escapable_handle_scope s,
+                                          napi_value v);
+
+
+// Methods to support error handling
+NODE_EXTERN void napi_throw(napi_env e, napi_value error);
+NODE_EXTERN void napi_throw_error(napi_env e, char* msg);
+NODE_EXTERN void napi_throw_type_error(napi_env e, char* msg);
+
+// These are particularly rough, needs more analysis
+typedef void (*napi_try_callback)(napi_env e, void* data);
+typedef void (*napi_catch_callback)(napi_env e, void* data);
+NODE_EXTERN bool napi_try_catch(napi_env e, napi_try_callback t,
+                                napi_catch_callback c, void* data);
+
+
+// Methods to provide node::Buffer functionality with napi types
+NODE_EXTERN napi_value napi_buffer_copy(napi_env e,
+                                        const char* data, uint32_t size);
+NODE_EXTERN bool napi_buffer_has_instance(napi_env e, napi_value v);
+NODE_EXTERN char* napi_buffer_data(napi_env e, napi_value v);
+NODE_EXTERN size_t napi_buffer_length(napi_env e, napi_value v);
+```
+
+# C++ Wrapper API
+
+The current view is that the API exported by the Node.js binary
+must be in C in order to achieve ABI stability. This [document]
+(http://www.oracle.com/technetwork/articles/servers-storage-dev/stablecplusplusabi-333927.html)
+outlines some of the challenges with respect to C++ and
+ABI stability.
+
+However, we also believe that a C++ wrapper is possible and
+would provide ease of use. The key element is that the C++ part must all
+be inlined and only use the external functions/types exported
+by the C API.  
```

Comment:
The problem with inlining is that there is no way to release bug fixes without everyone needing to recompile.